### ISSUE TITLE
dcache `num_alloc` incorrectly initialised?

### ISSUE THREAD

**Issue by daggaz** [archiecobbs/s3backer#89]
The `num_alloc` count in the `s3b_dcache` struct is incremented by `s3b_dcache_alloc_block` and decremented by `s3b_dcache_read_block`.

It is however not initialised by `s3b_dcache_init_free_list` when reading an existing cache file, so will be incorrect.

It also doesn't seem to be used anywhere?

**Comment by archiecobbs** [comment]
Good catch. I guess it's there for any users of the public API who want to know how full the cache is... but none exist yet.

Does this fix look right to you?
```
diff --git a/dcache.c b/dcache.c
index 5684f41..9b52916 100644
--- a/dcache.c
+++ b/dcache.c
@@ -686,6 +686,7 @@ s3b_dcache_init_free_list(struct s3b_dcache *priv, s3b_dcache_visit_t *visitor,
             }
         }
     }
+    priv->num_alloc = num_dslots_used;
 
     /* Reverse the free list so we allocate lower numbered slots first */
     for (i = 0; i < priv->free_list_len / 2; i++) {

```

**Comment by daggaz** [comment]
I'm not sure...

`num_dslots_used` is more like a high watermark of how far up the file we end up (max index)?

I think we'd need to increment the counter before we call the visitor?

**Comment by archiecobbs** [comment]
You're right, previous patch is wrong... `num_dslots_used` is the maximum in-use directory entry index plus one... this value is then used to verify that the file is not truncated (all in-use directory entries must have corresponding data).

This is more correct and also addresses your second point:
```
diff --git a/dcache.c b/dcache.c
index 5684f41..7cfbc34 100644
--- a/dcache.c
+++ b/dcache.c
@@ -679,6 +679,7 @@ s3b_dcache_init_free_list(struct s3b_dcache *priv, s3b_dcache_visit_t *visitor,
                 if ((r = s3b_dcache_push(priv, dslot)) != 0)
                     return r;
             } else {
+                priv->num_alloc++;
                 if (dslot + 1 > num_dslots_used)                    /* keep track of the number of dslots in use */
                     num_dslots_used = dslot + 1;
                 if (visitor != NULL && (r = (*visitor)(arg, dslot, entry->block_num, entry->md5)) != 0)
```

**Comment by archiecobbs** [comment]
Fixed in 8411409.

**Comment by daggaz** [comment]
:+1:
